### PR TITLE
feat:[CDS-74338]: fix for service override secret expression resolution in command step

### PR DESCRIPTION
**Pull request by sourabh-awashti**
## Describe your changes
Changes are regarding fix for command step where when we override service variable with secret value, then expression is not able to resolve, as we were using unresolved env variable expression which has been fixed.
Tests Done
1. Same secret variable present env , svc, and svc override for command step where  expression is getting resolved
2. variable is present only in env and expression is getting resolved
3. variable is present only in svc and expression is getting resolved

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50688)

### CODE REVIEW

**Comment by sourabh-awashti** [comment]
trigger uts

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/SshCommandStepHelper.java [hunk 1]
```diff
@@ -237,7 +237,6 @@ Map<String, String> getMergedEnvVariablesMap(
     finalEnvVariables = mergeEnvironmentVariables(evaluatedStageVariables, finalEnvVariables);
     finalEnvVariables = mergeEnvironmentVariables(evaluatedPipelineVariables, finalEnvVariables);
     finalEnvVariables = mergeEnvironmentVariables(getServiceVariables(ambiance), finalEnvVariables);
-    finalEnvVariables = mergeEnvironmentVariables(infrastructure.getEnvironment().getVariables(), finalEnvVariables);
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/SshCommandStepHelper.java:240]: can we write UT for this ?

Comment (sourabh-awashti) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/SshCommandStepHelper.java:240]: done

**Comment by sourabh-awashti** [comment]
trigger messagecheck

**Comment by sourabh-awashti** [comment]
trigger messagecheck

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/SshCommandStepHelper.java (hunk 1 above)
Comment (acasian-harness) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/SshCommandStepHelper.java:240]: @sourabh-awashti how the harness env variables are going to be resolved?

Comment (sourabh-awashti) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/SshCommandStepHelper.java:240]: we get the env variables in service variable outcome , hence we don;t need to add it again in variables map
cc: @yogesh-chauhan

**Comment by sourabh-awashti** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/SshCommandStepHelper.java` line 0 at r2](https://reviewable.io/reviews/harness/harness-core/50688#gh-1267760698:-N_hUjHm9WPum9mPyUVV:b-26jk5e) ([raw file](https://github.com/harness/harness-core/blob/003192434904733cb8e24b047551f84bc68bf0a1/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/ssh/SshCommandStepHelper.java#L0)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

The same question, let's say we are having only env variables, how does it be resolved? 
</blockquote></details>

even if we have only env variable , we have it in service outcome. While creating service outcome, we fetch env and service 
over ride variables, and then we create the map with all those variables. If there are multiple values present for same variable then we take the final value according the decided order but if we have single variable coming from env ,svc or svc override then we add it as it is.

**Review by acasian-harness — DISMISSED** [review]

**Review by ivan-83 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50688#-:-N_hZHpl5jD5RBdI-hnG:bnfp4nl)* status: 2 file(s) to review. 2. unresolved discussions.

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
*[`125-cd-nextgen/src/test/java/io/harness/cdng/ssh/SshCommandStepHelperTest.java` line 645 at r2](https://reviewable.io/reviews/harness/harness-core/50688#-N_hNG8sCU7Vi7q3gQTp:-N_haaMo6TRJZc4283_i:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/003192434904733cb8e24b047551f84bc68bf0a1/125-cd-nextgen/src/test/java/io/harness/cdng/ssh/SshCommandStepHelperTest.java#L645)):*
<details><summary><i>Previously, ivan-83 (Ivan Mijailovic) wrote…</i></summary><blockquote>

We have to covert the following test cases, 
- only service variables,
- only env variables
- only service override variables,
- service variable overridden by service override variable plus one env variable with a different name 
- service variable overridden by service override variable plus one env variable with the same name as service variable
</blockquote></details>

done

**Review by ivan-83 — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50688#-:-N_hefZA4MzZd3Qxl1jj:bnfp4nl)* status: 2 file(s) to review. 2. unresolved discussions. 1 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 1 of 1 files at r1, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50688)* status: 1. unresolved discussions.

**Comment by yogesh-chauhan** [comment]
trigger review

**Pull request merged**